processor/metricstransformprocessor/datapoint_aggregation.go

```diff
@@ -0,0 +1,259 @@
+// limitations under the License.
+
+package metricstransformprocessor
+
+import (
+	"fmt"
+	"math"
+	"math/rand"
+
+	metricspb "github.com/census-instrumentation/opencensus-proto/gen-go/metrics/v1"
+	"github.com/golang/protobuf/ptypes/timestamp"
+)
+
+// timeseriesGroupByLabelValues is a data structure for grouping timeseries that will be aggregated
+type timeseriesGroupByLabelValues struct {
+	timeseries  []*metricspb.TimeSeries
+	labelValues []*metricspb.LabelValue
+}
+
+// getLabelIdxs gets the indices of the labelSet labels' indices in the metric's descriptor's labels field
+// Returns the indices slice and a slice of the actual labels selected by this slice of indices
+func (mtp *metricsTransformProcessor) getLabelIdxs(metric *metricspb.Metric, labelSet map[string]bool) ([]int, []*metricspb.LabelKey) {
+	labelIdxs := make([]int, 0)
+	labels := make([]*metricspb.LabelKey, 0)
+	for idx, label := range metric.MetricDescriptor.LabelKeys {
+		_, ok := labelSet[label.Key]
+		if ok {
+			labelIdxs = append(labelIdxs, idx)
+			labels = append(labels, label)
+		}
+	}
+	return labelIdxs, labels
+}
+
+// aggregateTimeseriesGroups attempts to merge each group of timeseries in keyToTimeseriesMap through the specific calculation indicated by aggrType and dataType to construct the new timeseries with new label values indicated in keyToLabelValuesMap
+// Returns the new list of timeseries
+func (mtp *metricsTransformProcessor) aggregateTimeseriesGroups(keyToTimeseriesMap map[string]*timeseriesGroupByLabelValues, aggrType AggregationType, dataType metricspb.MetricDescriptor_Type) []*metricspb.TimeSeries {
+	newTimeSeriesList := make([]*metricspb.TimeSeries, len(keyToTimeseriesMap))
+	idxCounter := 0
+	for _, element := range keyToTimeseriesMap {
+		timestampToPoints, startTimestamp := mtp.analyzeTimeseriesForAggregation(element.timeseries)
+		newPoints := mtp.aggregatePoints(timestampToPoints, aggrType, dataType)
+		newSingleTimeSeries := &metricspb.TimeSeries{
+			StartTimestamp: startTimestamp,
+			LabelValues:    element.labelValues,
+			Points:         newPoints,
+		}
+		newTimeSeriesList[idxCounter] = newSingleTimeSeries
+		idxCounter++
+	}
+
+	return newTimeSeriesList
+}
+
+// analyzeTimeseriesForAggregation uses timestamp as an indicator to group aggregatable points in the group of timeseries, and determines the final value of the start timestamp for this groups of timeseries.
+// Returns a map that groups points by timestamps and the final start timestamp
+func (mtp *metricsTransformProcessor) analyzeTimeseriesForAggregation(timeseries []*metricspb.TimeSeries) (map[int64][]*metricspb.Point, *timestamp.Timestamp) {
+	var startTimestamp *timestamp.Timestamp
+	timestampToPoints := make(map[int64][]*metricspb.Point)
+	for _, ts := range timeseries {
+		if startTimestamp == nil || ts.StartTimestamp.Seconds < startTimestamp.Seconds || (ts.StartTimestamp.Seconds == startTimestamp.Seconds && ts.StartTimestamp.Nanos < startTimestamp.Nanos) {
```

Comment:
Could you add a comment for this logic?

Looks like it's assigning the timestamp as the start timestamp as long as it was before the start timestamp?

How do you know what cutoff timestamp to use?

Reply:
For this I am basically picking the start timestamp for the aggregated timeseries: I am picking the earliest start timestamp among the aggregated start timstamps because this should be the accurate start timestamp for this set of timeseries. Comments added.